### PR TITLE
[UNR-4522] Support replicating references to stably named dynamic subobjects of dynamic actors

### PR DESCRIPTION
**Pull request by improbable-valentyn**
#### Description
Native Unreal allows users to call `SetNetAddressable` on an actor component (or override `IsNameStableForNetworking` on a subobject) to allow replicating references to it even if it's not replicated, as long as it's created on server and client with the same name.
This PR adds support for this in the GDK, by adding logic in UObject pointer to UnrealObjectRef conversion for objects with `IsFullNameStableForNetworking() == false` and `IsNameStableForNetworking() == true`.

#### Release note
- Fixed an issue with replicating references to stably named dynamically added subobjects of dynamic actors.

#### Tests
Added a functional test that validates that passing a reference to a partially stably named object into an RPC works correctly.

#### Primary reviewers
@m-samiec

### CODE REVIEW

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialPackageMapClient.cpp [hunk 1]
```diff
@@ -728,9 +728,15 @@ FNetworkGUID FSpatialNetGUIDCache::GetNetGUIDFromUnrealObjectRefInternal(const F
...
 
 		// Recursively resolve the outers for this object in order to ensure that the package can be loaded
-		if (ObjectRef.Outer.IsSet())
+		if (ObjectRef.Outer.IsSet() && ObjectRef.Outer.GetValue() != FUnrealObjectRef::NULL_OBJECT_REF)
 		{
 			OuterGUID = GetNetGUIDFromUnrealObjectRef(ObjectRef.Outer.GetValue());
+
+			if (!OuterGUID.IsValid())
+			{
+				// Couldn't resolve the outer, most likely because it's a dynamic actor that we haven't received yet.
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialPackageMapClient.cpp:737]: Should this be an error?

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialPackageMapClient.cpp [hunk 2]
```diff
@@ -771,10 +777,12 @@ void FSpatialNetGUIDCache::NetworkRemapObjectRefPaths(FUnrealObjectRef& ObjectRe
 
 void FSpatialNetGUIDCache::UnregisterActorObjectRefOnly(const FUnrealObjectRef& ObjectRef)
 {
-	FNetworkGUID& NetGUID = UnrealObjectRefToNetGUID.FindChecked(ObjectRef);
-	// Remove ObjectRef first so the reference above isn't destroyed
-	NetGUIDToUnrealObjectRef.Remove(NetGUID);
-	UnrealObjectRefToNetGUID.Remove(ObjectRef);
+	if (FNetworkGUID* NetGUID = UnrealObjectRefToNetGUID.Find(ObjectRef))
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialPackageMapClient.cpp:780]: Seems we only use this for stably named actors currently, under what circumstances might the object ref not be registered?

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/PartiallyStablePathTest/PartiallyStablePathActor.h [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+{
+	GENERATED_BODY()
+
+public:
+	APartiallyStablePathActor() {}
+
+	virtual void BeginPlay() override;
+
+	UPROPERTY()
+	UStaticMeshComponent* Component;
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/PartiallyStablePathTest/PartiallyStablePathActor.h:23]: nit: DynamicComponent

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/PartiallyStablePathTest/PartiallyStablePathTest.cpp [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+ * Partially Stable Path Test
+ *
+ * This test aims to verify that references to dynamically added subobjects with a partially
+ * stable path are replicated correctly. A partially stable path means an object has a stable
+ * path relative to its outer, but its full path is not stable, meaning one of its outers is
+ * a dynamic actor.
+ *
+ * The test includes 1 server and 1 client.
+ * The flow is as follows:
+ * - Setup:
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/PartiallyStablePathTest/PartiallyStablePathTest.cpp:24]: Empty flow

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/PartiallyStablePathTest/PartiallyStablePathTest.cpp [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+}
+
+void APartiallyStablePathTest::PrepareTest()
+{
+	Super::PrepareTest();
+
+	AddStep(TEXT("Spawn the test actor"), FWorkerDefinition::Server(1), nullptr, [this]() {
+		// Verify the pawn has the right class
+		ASpatialFunctionalTestFlowController* FlowController = GetFlowController(ESpatialFunctionalTestWorkerType::Client, 1);
+		APlayerController* PlayerController = Cast<APlayerController>(FlowController->GetOwner());
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/PartiallyStablePathTest/PartiallyStablePathTest.cpp:42]: CastChecked for all these?

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/PartiallyStablePathTest/PartiallyStablePathTest.cpp [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+		},
+		[this]() {
+			TArray<AActor*> TestActors;
+			UGameplayStatics::GetAllActorsOfClass(GetWorld(), APartiallyStablePathActor::StaticClass(), TestActors);
+
+			AssertEqual_Int(TestActors.Num(), 1, TEXT("Number of test actors on the client"));
+			APartiallyStablePathActor* TestActor = Cast<APartiallyStablePathActor>(TestActors[0]);
+
+			APlayerController* PlayerController = Cast<APlayerController>(GetLocalFlowController()->GetOwner());
+			APartiallyStablePathPawn* Pawn = Cast<APartiallyStablePathPawn>(PlayerController->GetPawn());
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/PartiallyStablePathTest/PartiallyStablePathTest.cpp:71]: Seems unlikely the Pawn hasn't arrived yet, but safer to check

File: SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/PartiallyStablePathTest/PartiallyStablePathTest.h [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+{
+	GENERATED_BODY()
+
+public:
+	APartiallyStablePathTest();
+
+	virtual void PrepareTest() override;
+
+private:
+	APartiallyStablePathPawn* Pawn;
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/PartiallyStablePathTest/PartiallyStablePathTest.h:25]: UPROPERTY

**Review by m-samiec — APPROVED** [review]
Nice

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialPackageMapClient.cpp (hunk 1 above)
Comment (improbable-valentyn) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialPackageMapClient.cpp:737]: No, returning an empty GUID here will be treated as a regular unresolved reference.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialPackageMapClient.cpp (hunk 2 above)
Comment (improbable-valentyn) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialPackageMapClient.cpp:780]: This change is meant to catch an edge case where an object claims it's fully stably named (by overriding `IsFullNameStableForNetworking`) but actually isn't. It's not really a valid use case but still better not to crash.

**Pull request merged**